### PR TITLE
Support easy run daemon mode for Linux/MacOS

### LINKED ISSUE #4237: Easy daemon mode for ckb run
#### Is your feature request related to a problem? Please describe.
There is no simple approach to run a ckb node in daemon or background mode. Although there is an introduction to [start a ckb service](https://github.com/nervosnetwork/ckb/tree/develop/devtools/init), it's still too complex for average users.

#### Describe the solution you'd like
Similar to bitcoind, if ckb run provide a simple switch like `ckb run --daemon`, I believe more nodes will appear.

### PR DESCRIPTION
**Pull request by chenyukang**
### What problem does this PR solve?

Issue Number: close #4237

Problem Summary:

### What is changed and how it works?

- use crate `daemonize` for supporting daemon 
- use crate `colored` for better terminal UX
- Note: `new_global_runtime` need to be invoked after daemonize success

Command line:
- `ckb run --daemon` - run ckb in daemon 
- `ckb daemon --check` - check the status of service
- `ckb daemon --stop` - stop the daemon process

What's Changed:

### Related changes

- PR to update `owner/repo`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code ci-runs-only: [ quick_checks,linters ]

Side effects

- Performance regression
- Breaking backward compatibility

### Release note 

```release-note
Title Only: Include only the PR title in the release note.
```

### CODE REVIEW

File: ckb-bin/src/subcommand/daemon.rs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            }
+            _ => {
+                eprintln!("{:>10} : {:<12}", name, "not running".red());
+            }
+        }
+    } else if args.stop {
+        kill_process(pid_file, "ckb")?;
+        fs::remove_file(pid_file).map_err(|_| ExitCode::Failure)?;
+    } else {
+        unimplemented!()
```
Comment (quake) [inline ckb-bin/src/subcommand/daemon.rs:25]: should print help message instead of panic

File: ckb-bin/src/subcommand/daemon.rs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    let pid_str = fs::read_to_string(pid_file).map_err(|_| ExitCode::Failure)?;
+    let pid = pid_str
+        .trim()
+        .parse::<i32>()
+        .map_err(|_| ExitCode::Failure)?;
+    eprintln!("kill {} process {} ...", name, pid.to_string().red());
+    // Send a SIGTERM signal to the process
+    let _ = kill(Pid::from_raw(pid), Some(Signal::SIGTERM)).map_err(|_| ExitCode::Failure);
+    // sleep 3 seconds and check if the process is still running
+    std::thread::sleep(std::time::Duration::from_secs(3));
```
Comment (eval-exec) [inline ckb-bin/src/subcommand/daemon.rs:58]: Is a 3 second delay sufficient before shutting down ckb?

File: ckb-bin/src/subcommand/daemon.rs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    match check_process(pid_file) {
+        Ok(_) => kill(Pid::from_raw(pid), Some(Signal::SIGKILL)).map_err(|_| ExitCode::Failure),
```
Comment (eval-exec) [inline ckb-bin/src/subcommand/daemon.rs:60]: I'm concerned that using `SIGKILL` may not be appropriate.

File: ckb-bin/src/lib.rs [hunk 4]
```diff
@@ -56,8 +61,64 @@ pub fn run_app(version: Version) -> Result<(), ExitCode> {
...
+    cmd: &str,
+    matches: &ArgMatches,
+) -> Result<(), ExitCode> {
+    eprintln!("starting CKB in daemon mode ...");
+    eprintln!("check status : `{}`", "ckb daemon --check".green());
+    eprintln!("stop daemon  : `{}`", "ckb daemon --stop".yellow());
+
+    assert!(matches!(cmd, cli::CMD_RUN));
+    let root_dir = Setup::root_dir_from_matches(matches)?;
+    let daemon_dir = root_dir.join("data/daemon");
```
Comment (quake) [inline ckb-bin/src/lib.rs:85]: We usually create the pid file in the tmp directory
```suggestion
    let daemon_dir = root_dir.join("tmp/daemon");
```

File: ckb-bin/src/subcommand/daemon.rs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    if check_process(pid_file).is_err() {
+        eprintln!("{} is not running", name);
+        return Ok(());
+    }
+    let pid_str = fs::read_to_string(pid_file).map_err(|_| ExitCode::Failure)?;
+    let pid = pid_str
+        .trim()
+        .parse::<i32>()
+        .map_err(|_| ExitCode::Failure)?;
+    eprintln!("kill {} process {} ...", name, pid.to_string().red());
```
Comment (quake) [inline ckb-bin/src/subcommand/daemon.rs:54]:
```suggestion
    eprintln!("stopping ckb daemon service {} ...", name, pid.to_string().red());
```

File: ckb-bin/src/subcommand/daemon.rs [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+use std::path::PathBuf;
+
+pub fn daemon(args: DaemonArgs) -> Result<(), ExitCode> {
+    let name = "ckb";
+    let pid_file = &args.pid_file;
+    if args.check {
+        // find the pid file and check if the process is running
+        match check_process(pid_file) {
+            Ok(pid) => {
+                eprintln!("{:>10} : {:<12} pid - {}", name, "running".green(), pid);
```
Comment (quake) [inline ckb-bin/src/subcommand/daemon.rs:15]:
```suggestion
                eprintln!("{:>10} : {:<12} pid - {}", "ckb daemon service is running".green(), pid);
```

File: ckb-bin/src/subcommand/daemon.rs (hunk 3 above)
Comment (chenyukang) [reply ckb-bin/src/subcommand/daemon.rs:60]: This is the last method after waiting for too long.
Any better solution for it?

File: ckb-bin/src/subcommand/daemon.rs (hunk 1 above)
Comment (chenyukang) [reply ckb-bin/src/subcommand/daemon.rs:25]: invalid option will be terminated with help message by the caller of this function.
I think it also OK to remove this assertion.

File: ckb-bin/src/lib.rs [hunk 7]
```diff
@@ -56,8 +61,64 @@ pub fn run_app(version: Version) -> Result<(), ExitCode> {
...
 
+    if run_deamon(cmd, matches) {
+        run_app_in_daemon(version, bin_name, cmd, matches)
+    } else {
+        debug!("ckb version: {}", version);
+        run_app_inner(version, bin_name, cmd, matches)
+    }
+}
+
+fn run_app_in_daemon(
```
[the remark this answers is not in the document]
Comment (chenyukang) [reply ckb-bin/src/lib.rs:73]: My first version is add daemon related code in run,
then I found some issues, the 
```rust
    let (mut handle, mut handle_stop_rx, _runtime) = new_global_runtime();
```
must be invoked in the child process.
So I moved the new code to a up level.

File: ckb-bin/src/lib.rs (hunk 4 above)
Comment (chenyukang) [reply ckb-bin/src/lib.rs:85]: @zhangsoledad suggestted add a new `daemon` sub-dirctory, 
which we can also store other things from daemon in future, and a fixed path may be better?
I think both are ok, if we don't cleanup tmp.
And we also won't create `daemon` sub-directory if user don't start ckb as daemon service.

**Comment by chenyukang** [comment]
> a sentry log is printed when I run with `ckb daemon --check`, I think we should remove it:
> 
> ```
>        ckb : running      pid - 81683
> 2023-11-29 02:01:44.763 +00:00 main INFO sentry  sentry is disabled
> ```

Removed.

**Review by zhangsoledad — APPROVED** [review]

**Review by eval-exec — APPROVED** [review]

**Review by quake — APPROVED** [review]

**Pull request merged**